pkg/kv/kvserver/logstore/sideload_disk.go

```diff
@@ -102,15 +99,28 @@ func (ss *DiskSideloadStorage) Put(
 		} else if !oserror.IsNotExist(err) {
 			return err
 		}
-		// createDir() ensures ss.dir exists but will not create any subdirectories
-		// within ss.dir because filename() does not make subdirectories in ss.dir.
-		if err := ss.createDir(); err != nil {
+		// Ensure that ss.dir exists. The filename() is placed directly in ss.dir,
+		// so the next loop iteration should succeed.
+		if err := mkdirAllAndSyncParents(ss.eng, ss.path[0], ss.path[1:], os.ModePerm); err != nil {
```

Comment:
NB: this will sync parents only for the newly created directories. For instance:

- if `auxiliary/sideloading/r1XXX/r1234` already exists then there will be no sync
- if `auxiliary/sideloading/r2XXX` exists and we add `auxiliary/sideloading/r2XXX/r2222`, then only the `r2XXX` folder is synced.

This leaves a window of vulnerability in existing clusters who already have sideloaded files and did not sync the folders before. Only **newly** created sideloaded storages are strongly synced.

We could make this stronger for the existing clusters too, by ensuring that every pre-existing folder is synced at least once. Some ways of doing this:

1. add a migration that traverses and syncs the sideloaded storage at startup
2. or, lazily force a sync once per directory (but we would need the overhead of tracking which ones are [not] synced)
3. or, call something like [syncfs](https://linux.die.net/man/2/syncfs) on startup.

However, in practice, only recently created directories are susceptible to "disappearing" after a power outage. I think it's fine to accept this little risk, and not introduce any migration complexity. The sideloaded storage directories will be fully synced when a binary with this patch is started, **and** a bit of time has passed so that the FS has flushed all the pre-existing dirs. After that, the storage is always synced. I.e. time is our "migration".

Reply:
I don't think we need to consider existing clusters here. Since the Raft state and log are stored in the same Pebble instance, I believe applied state changes (i.e. the SST hardlink) will be synced on the next Raft append, which will happen shortly after a restart with the new binary. Does that sound right?

Reply:
Yeah, the SSTs are hard linked into Pebble on apply stage when doing the ingestion (which also syncs the necessary parts of the FS): https://github.com/cockroachdb/cockroach/blob/14733120c08ba1a0a050bb5442c3311090206566/pkg/kv/kvserver/replica_proposal.go#L724

But, I think, the references to these files from the log's directory may still remain un-synced. So if we restart after a crash, it's possible that the entries are "gone" from the log. Raft may try to access these log entries (e.g. to apply them again, or to catch up other followers).

Reply:
Also, this argument only applies to SSTs that were applied. The unapplied ones are more vulnerable.